Updates glossary with many terms and links. (#79)
NUbots/NUbook: src/book/03-guides/04-general/03-glossary.mdx

[**Euler Angles**](https://en.wikipedia.org/wiki/Euler_angles): Roll, Pitch, and Yaw - they describe rotation for three orthogonal axes.

[**Inertial Measurement Unit (IMU)**](https://en.wikipedia.org/wiki/Inertial_measurement_unit): A device which measures acceleration with accelerometers and rotation using gyroscopes.

[**Inverse-Kinematics**](https://en.wikipedia.org/wiki/Inverse_kinematics): Calculation of joint parameters for the end of a kinematic chain relative to the start of the kinematic chain.

[**Kalman Filter**](https://en.wikipedia.org/wiki/Kalman_filter): Control theoretic technique for estimating variables by estimating a joint probability distribution over the variables for each timeframe.

[**Kinematics**](https://en.wikipedia.org/wiki/Kinematics): Mathematics which describes motion, which is used to calculate trajectories and servo outputs to control robots.

[**Lambda Expression**](https://en.cppreference.com/w/cpp/language/lambda): A way of defining an anonymous function in C++.

[**Localisation**](https://en.wikipedia.org/wiki/Robot_navigation): The system the robot uses to work out its position relative to its environment and objects within it.

> [**Odometry**](https://en.wikipedia.org/wiki/Odometry): Estimation of change in position based on sensor commands and data.

Author: Don't we take sensor outputs into account for odometry? I had assumed you would

[**NUClear Roles**](https://github.com/Fastcode/NUClearRoles): The build and messaging system used to manage the construction and messaging of NUClear for the NUbots codebase.

**NUFSR**: A system designed to detect if a NUgus' foot is on the ground, by use of force sensitive resistors (FSRs).

[**NUgan**](https://github.com/NUbots/NUgan): A custom [cycle-gan](https://junyanz.github.io/CycleGAN/) implementation to generate more realistic images for training the vision networks.

[**NUgus**](https://nubook.nubots.net/system/hardware/overview#igus-modifications): NUbots robots as currently used in the RoboCup humanoid league. They are modified [igus Humanoid Open Platform](https://arxiv.org/abs/1809.11110) robots, which we 3D-printed.

[**NUpbr**](https://github.com/NUbots/NUpbr): The software NUbots uses to generate Physically Based Rendered football fields, used to train vision networks.

[**NUsight**](https://github.com/NUbots/NUsight2): NUbots' visualisation software, ran through a browser on an external computer.

> **NUsense**: A secret project for sensor IO.

Reviewer: ```suggestion
**NUSense**: A secret project for sensor IO.
```

**Old Walk Engine**: A previously used walk engine, originally optimised using genetic algorithms.

**Quintic Walk**: Open loop walk engine which uses quintic splines to create trajectories. Created by Bit-Bots, based off code from team Rhoban, then ported to run with NUClear.

**ScriptTuner**: A NUbots program for manually creating motion targets for movement scripts.

**ScriptRunner**: The program which runs movement scripts on the robot, such as `Zombie.yaml` which is used to test servo installations.

**Static Motion**: We tend to refer to motion which is fully controlled by scripts which do not change at run-time as _Static Motion_.

> **Subsumption**: TODO!

Reviewer: ```suggestion
**Subsumption**: The priority system that determines if a neutron gets access to the servos it wants to control.
```
Might need rewording. But something like this.

Reviewer 2: We should probably briefly describe what a neutron is here in parenthesis, or add a new term in the glossary and link to that, or link to a definition somewhere in the NUClear docs.

[**VisualMesh**](https://arxiv.org/abs/1807.08405): A computer vision technique NUbots uses which overlays a mesh over the image input to more efficiently sample it and look for objects. Source can be found [here](https://github.com/Fastcode/VisualMesh).

### General

[**Affine Transform**](https://en.wikipedia.org/wiki/Affine_transformation): A transform which keeps straight lines straight, and parallel lines parallel, without preserving distances or angles between lines otherwise.

> [**Closed Loop Control**](https://en.wikipedia.org/wiki/Closed-loop_transfer_function): Control which uses output from the thing being used as input for controlling it.

Reviewer: ```suggestion
[**Closed Loop Control**](https://en.wikipedia.org/wiki/Closed-loop_transfer_function): Control action which uses output from the system as input.
```

[**Kinematics**](https://en.wikipedia.org/wiki/Kinematics): Mathematics which describes motion, which is used to calculate trajectories and servo outputs to control robots.

[**Lambda Expression**](https://en.cppreference.com/w/cpp/language/lambda): A way of defining an anonymous function in C++.

[**Localisation**](https://en.wikipedia.org/wiki/Robot_navigation): The system the robot uses to work out its position relative to its environment and objects within it.

[**Machine Learning**](https://en.wikipedia.org/wiki/Machine_learning): Systems and models which improve their outputs by exposure to inputs.

[**Mahony Filter**](https://nitinjsanket.github.io/tutorials/attitudeest/mahony): A complementary filter which estimates attitude, angle, and orientation using IMU data.

[**Motion Capture (mocap)**](https://en.wikipedia.org/wiki/Motion_capture): The recording of movements of moving objects, which keeps track of specific points on the object as it moves.

[**Odometry**](https://en.wikipedia.org/wiki/Odometry): Estimation of change in position and orientation based on sensor inputs.

> [**Open Loop Controller**](https://en.wikipedia.org/wiki/Open-loop_controller): A controller whose actions aren't affected by feedback from its actions.

Reviewer: ```suggestion
[**Open Loop Controller**](https://en.wikipedia.org/wiki/Open-loop_controller): Control action that does not use output from the system as input.
```

[**CM740**](<http://support.robotis.com/en/product/robotis-op2/sub_controller(cm-740).htm>): A _Robotis_ sub-controller which gives the commands to the servos.

[**CMake**](https://cmake.org/): An open-source build system generator NUbots uses to configure and build the main codebase.

[**Darwin**](https://robots.ieee.org/robots/darwin/): A _Robotis_ robot platform which was the platform NUbots used before making the NUgus.

[**Docker**](https://www.docker.com/): A software service for developing containers which enable standardised cross-platform development consistency.

[**Dockerhub**](https://hub.docker.com/): A hosting service for docker container images.

[**Dynamixel**](https://www.robotis.us/dynamixel/): The _Robotis_ servo models which NUgus uses.

[**igus**](https://www.igus.com.au/): The creators of the robotics platform NUgus is based on.

[**igus Humanoid Open Platform**](https://arxiv.org/abs/1809.11110): The 3d printed base design which the NUgus is based on.

[**Eigen**](https://eigen.tuxfamily.org/): The C++ linear algebra library NUbots uses. Documentation can be found [here](https://eigen.tuxfamily.org/dox/index.html).

> [**Gazebo**](https://gazebosim.org/): A simulation software used which was used to optimise the "OldWalkEngine" with a genetic algorithm.

Reviewer: ```suggestion
[**Gazebo**](https://gazebosim.org/): A simulation software that is used to optimise motions with a genetic algorithm.
```

Author: I don't agree with this one. This makes it seem like that's the point of gazebo, but gazebo is a lot more than just that project - especially this year with RoboCup

[**OpenCL**](https://www.khronos.org/opencl/): Open Computing Language is a framework NUbots uses to create programs which can run on graphics hardware.

[**Protobuf**](https://developers.google.com/protocol-buffers): Protocol Buffers (Protobuf) is a method of serializing structured data, created by Google. Protobuf messages are the means of passing messages between NUClear reactors.

[**Rhoban**](https://www.rhoban-project.fr/): The RoboCup team from Bordeaux University, who created the original quintic walk adapted by Bit-Bots, and ported to the NUbots codebase.

[**RoboCup**](https://www.robocup.org/): The robotic football world cup, held each year in different cities around the world.

[**RoboCup Symposium**](https://www.robocup.org/symposium): The research conference which is held after RoboCup each year, featuring research from the teams competing.

[**Robotis**](https://www.robotis.us/): The Korean robotics manufacturer which makes the servos that NUgus uses.

[**Simulink**](https://www.mathworks.com/products/simulink.html): A MATLAB-based graphical simulation software.

> [**Simscape**](https://www.mathworks.com/products/simscape.html): A MATLAB-based simulation tool for creating models of physical systems in the Simulink environment.

Reviewer: ```suggestion
[**Simscape**](https://www.mathworks.com/products/simscape.html): A MATLAB-based simulation tool for creating models of physical systems in the Simulink environment. Also the name of the GitHub repository that contains the solution that models and simulates the NUgus robot using Simulink and Simscape.
```
